sdk/core/System.ClientModel/src/Message/MultipartContent.cs

```diff
@@ -0,0 +1,295 @@
+        if (string.IsNullOrWhiteSpace(boundary))
+        {
+            throw new ArgumentException("Value cannot be null or empty.", boundary);
+        }
+
+        // RFC 2046 Section 5.1.1
+
+        if (boundary.Length > 70)
+        {
+            throw new ArgumentOutOfRangeException(nameof(boundary), boundary, $"The field cannot be longer than {70} characters.");
+        }
+
+        if (boundary.EndsWith(" ", StringComparison.InvariantCultureIgnoreCase))
+        {
+            throw new ArgumentException($"The format of value '{boundary}' is invalid. '{boundary}' cannot end in a space", nameof(boundary));
+        }
+
+        foreach (char ch in boundary)
+        {
+            const string AllowedMarks = @"'()+_,-./:=? ";
+            if (!(('0' <= ch && ch <= '9') || // Digit.
+                    ('a' <= ch && ch <= 'z') || // alpha.
+                    ('A' <= ch && ch <= 'Z') || // ALPHA.
+                    AllowedMarks.Contains(char.ToString(ch)))) // Marks.
+            {
+                throw new ArgumentException($"The format of value '{boundary}' is invalid.", nameof(boundary));
+            }
+        }
+    }
+
+    /// <summary>
+    /// Releases the unmanaged resources and disposes of the managed resources used by the <see cref="MultipartContent"/>.
+    /// </summary>
+    public void Dispose()
+    {
+    }
+
+    #region Subpart
+
+    /// <summary>
+    /// A helper class representing a subpart of a multipart content.
+    /// </summary>
+    private class MultipartContentSubpart
+    {
+        /// <summary>
+        /// The <see cref="BinaryContent"/> that represents the content of this subpart.
+        /// </summary>
+        public BinaryContent Content { get; }
+
+        /// <summary>
+        /// The headers for this subpart.
+        /// </summary>
+        public Dictionary<string, string> Headers { get; }
+
+        public MultipartContentSubpart(BinaryContent content, Dictionary<string, string> headers)
+        {
+            Content = content;
+            Headers = headers;
+        }
+
+        /// <summary>
+        /// Writes the <see cref="Headers"/> and <see cref="Content"/> of this subpart to the provided <see cref="Stream"/>.
+        /// </summary>
+        /// <param name="stream">The <see cref="Stream"/> to write to.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> to signal cancellation of the write operation.</param>
+        public void WriteTo(Stream stream, CancellationToken cancellationToken)
+        {
+            // Write headers on a new line
+            stream.Write(_crLfBytes, 0, _crLfBytes.Length);
+
+            // Write the headers to stream
+            foreach (var header in Headers)
+            {
+                string headerString = $"{header.Key}{ColonSpace}{header.Value}{CrLf}";
+                byte[] headerBytes = Encoding.UTF8.GetBytes(headerString);
+                stream.Write(headerBytes, 0, headerBytes.Length);
```

Comment:
Is there any wisdom in checking the cancellation token on each loop iteration, or do we expect these values to be so small it shouldn't matter?

Reply:
updated